utils/swlite-authorize.1
.\" swlite-authorize - manage sipwitchqt user authorizations in sqlite backend
.\" Copyright (C) 2017-2018 Tycho Softworks
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH swlite-authorize "1" "December 2017" "SipWitchQt" "Tycho Softworks"
.SH NAME
swlite-authorize \- manage sipwitchqt user authorizations in sqlite backend
.SH SYNOPSIS
.B swlite-authorize
.RI [ options ]
.I user
.br
.B swlite-authorize
.br
.SH DESCRIPTION
The purpose of swlite-authorize is to manage authorization group records for a backend 
sqlite sipwitchqt database.  This includes both creating and removing (or suspending)
authorized users, and also setting the SIP digest password.

If used with no options the current authorization realm is shown.  If a userid is given
then the password to use will be prompted on the console.  Otherwise optios can be
used to set digest mode, and whether to suspend or drop a user.
.SH OPTIONS
.TP
.B \-\-sha256
Use a SHA-256 digest when setting a password (default is MD5).
.TP
.B \-\-sha512
Use a SHA-512 digest when setting a password (default is MD5).
.TP
.B \-\-create
Create (add) the authorization if it does not exist.
.TP
.B \-\-drop
Drop (remove) the authorized user and associated extension record.  Any active calls or
registrations should also be terminated.
.TP
.B \-\-local
Local device group authorization.
.TP
.BI \-\-password value
Force setting of password.  Useful for script driven db setup.
.TP
.B \-\-remote
Remote accessible user authorization.
.TP
.B \-\-suspend
Suspend the users account.  Any active calls and registrations should also be cancelled.
.TP
.B \-\-team
Team group authorization.
.SH AUTHOR
.B swlite-authorize
was written by David Sugar <tychosoft@gmail.com>.
.SH "REPORTING BUGS"
Report bugs to tychosoft@gmail.com.
.SH COPYRIGHT
Copyright \(co 2017 Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



CHANGELOG
v0.2.7
* Improved packaging support
* Updated copyright notices
* Utils use stand-alone to init mysql servers
* De-emphasizing sqlite beyond testing builds

v0.2.6
* Simplified expiration handling
* Re-arranged font selection
* Introduced desktop & audio notify setting
* Integrated login auto start behavior
* Speed dial ui support
* Call coverage setting support
* Call forwarding setting support
* Delay ringing option supported
* Per user and global auto answer settings
* Start of integrated voice mail
* Device keys introduced

v0.2.5
* Customization at build time
* Topic management in sessions
* Policy setting for main public room

v0.2.4
* Session/state based toolbar management
* Online status of all extensions shown
* Membership management for groups
* Server system notification messaging
* Delete authorizations supported

v0.2.3
* Schema changes
* Message expiration

v0.2.2
* Change password support
* Simple markdown support
* Zeroconfg support
* Message sync for existing extensions
* Functional testing support
* Improved adduser ui
* Expiration marking on client

v0.2.1
* Fix for crash sending offline messages
* Visual styling improvements and optimizations

V0.2.0
* IPL loading of tables, such as for external mysql db
* User classification of priviledges
* Ability to edit extension names from profile
* Clickable and highlighted urls in messages
* Offline message sync
* Cleaner db handling for login/logout and multiple dbs
* Client import/export
* Font picker for message view
* Admin can create users from client
* Phonebook expanded
* New message indicators

v0.1.4
* I want to send a message to someone
* I Want to receive messages
* I want to search my messages
* I want to logout, and back in
* I want to know what extensions I can login as
* I want the ui to become available in language of my choice
* I want to know what devices are on my extension
* I want to autostart when I login on linux
* sipwitchqt message laying and rendering engine introduced
* email and public uri properties added to contacts

v0.1.3
* Roster processing on client and server
* Search processing on sessions screen
* click feedback, doubleclick contacts
* vibrant and light icon schemes

v0.1.2
* Start of sessions page and status

v0.1.1
* udp for eXosip2 messaging support
* automatic authentication integrated
* Start of core messaging code
* Replacing json with xdp
* Start of contacts (phonebook) page

v0.1.0
* Core authentication complete 
* Core registration with expiration
* Extended registration support for sipwitchqt clients

v0.0.2
* More complete test harness debug config
* Reset and dialogs for client
* Release packaging for primary platforms

v0.0.1
* Initial development testing release



utils/swlite-authorize.rb
#!/usr/bin/env ruby
# Copyright (C) 2017-2018 Tycho Softworks
#
# This file is free software; as a special exception the author gives
# unlimited permission to copy and/or distribute it, with or without
# modifications, as long as this notice is preserved.

['optparse', 'io/console', 'sqlite3', 'digest'].each {|mod| require mod}

if STDIN.respond_to?(:noecho)
  def get_password(prompt="Password: ")
    print prompt
    STDIN.noecho(&:gets).chomp
  end
else
  def get_password(prompt="Password: ")
    `read -s -p "#{prompt}" password; echo $password`.chomp
  end
end

access = 'LOCAL'
digest = 'MD5'
number = -1
minext = 100
maxext = 699
user = nil
password = nil
mode = 'USER'
create = false
domain = nil
dbname = '/var/lib/sipwitchqt/local.db'
dbname = '../testdata/local.db' if File.writable?('../testdata/local.db')
dbname = '../userdata/local.db' if File.writable?('../userdata/local.db')
abort("*** swlite-authorize: no database") unless File.writable?(dbname)

OptionParser.new do |opts|
  opts.banner = 'Usage: swlite-authorize [id]'

  opts.on('-c', '--create', 'deauthorize and remove user') do
    create = true
  end

  opts.on('-d', '--drop', 'deauthorize and remove user') do
    digest = 'DROP'
  end

  opts.on('-l', '--local', 'local device group') do
    mode = 'DEVICE'
    access = 'LOCAL'
  end

  opts.on('-p', '--password [SECRET]', 'force password set') do |secret|
    password = secret
  end

  opts.on('-r', '--remote', 'remote user access') do
    access = 'REMOTE'
  end

  opts.on('-s', '--suspend', 'suspend user') do
    digest = 'NONE'
  end

  opts.on('-t', '--team', 'team authorization') do
    mode = 'TEAM'
    access = 'TEAM'
  end

  opts.on('-2', '--sha256', 'use sha-256') do
    digest = 'SHA-256'
  end

  opts.on('-5', '--sha512', 'use sha-512') do
    digest = 'SHA-512'
  end

  opts.on_tail('-h', '--help', 'Show this message') do
    puts opts
    exit
  end
end.parse!
abort(opts.banner) if(ARGV.size > 1)
user = ARGV[0]

begin 
  db = SQLite3::Database.open dbname
  db.execute("SELECT realm, dialing FROM Config") do |row|
    abort("*** swlite-authorize: multiple config entries") unless domain == nil
    domain = row[0]
    dialing = row[1]
    case dialing
    when 'STD3'
      minext = 100
      maxext = 699
    else
      abort("*** swlite-authorize: #{dialing}: invalid dialing plan used")
    end
  end
rescue SQLite3::SQLException
  abort("*** swlite-authorize: no switch table")
end

print "Realm #{domain}\n"
exit if user == nil

begin
  type = 'NONE'
  db.execute("SELECT name, type, access FROM Authorize WHERE name='#{user}'") do |row|
    type = row[1]
    access = row[2]
  end
  if type == 'NONE'
    abort("*** swlite-authorize: #{user}: no such authorization") if create != true or digest == 'DROP'
    db.execute("INSERT INTO Authorize (name, type, realm, access) VALUES('#{user}','#{mode}','#{domain}','#{access}')")
    type = mode
    print "Created #{user}\n"
  end
  case digest
  when 'NONE'
    print "Suspending authorization #{user}\n"
    db.execute("UPDATE Authorize set digest='NONE', secret='', realm='#{domain}' WHERE name='#{user}'")
  when 'DROP'
    case type
    when 'USER', 'DEVICE', 'TEAM'
      print "Dropping authorization #{user}\n"
      db.execute("DELETE FROM Authorize WHERE name='#{user}'")
    else
      abort("*** swlite-authorize: #{user}: cannot drop authorization as #{type}")
    end
  else
    case access
    when 'LOCAL', 'REMOTE', 'TEAM'
      print "Changing password for #{user}\n"
      if password == nil
        pass1 = get_password
        print "\n"
        exit unless pass1.size > 0
        pass2 = get_password "Retype Password: "
        print "\n"
      else
        pass1 = pass2 = password
      end
      exit unless pass2.size > 0
      abort("*** swlite-authorize: passwords do not match") unless pass1 == pass2
      case digest
      when 'SHA-256'
        secret = Digest::SHA256.hexdigest "#{user}:#{domain}:#{pass1}"
      when 'SHA-512'
        secret = Digest::SHA512.hexdigest "#{user}:#{domain}:#{pass1}"
      else
        secret = Digest::MD5.hexdigest "#{user}:#{domain}:#{pass1}"
      end
      db.execute("UPDATE Authorize set digest='#{digest}', secret='#{secret}', realm='#{domain}' WHERE name='#{user}'")
    else
      abort("*** swlite-authorize: #{user}: not able to authorize as #{access}")
    end
  end
rescue SQLite3::BusyException
  abort("*** swlite-authorize: database busy; sipwitchqt active")
rescue SQLite3::SQLException
  abort("*** swlite-authorize: database error")
rescue Interrupt
  abort("^C")
  exit
end

